### ISSUE TITLE
test 6, "allocate_as_barier_proc", failing

### ISSUE THREAD

**Issue by zbeekman** [sourceryinstitute/opencoarrays#28]
See #27 for some context/backstory

Is there a known issue with test 6 "allocate_as_brier_proc" when using OpenMPI and GFortran 5.2? I know from the [known issues](https://github.com/sourceryinstitute/opencoarrays/blob/master/STATUS.md#known-issues) that "co_reduce" won't work until GFortran 5.3/6.1 but what's the story with this other test?

**Comment by zbeekman** [comment]
This question is intended primarily for @afanfa, I believe

**Comment by zbeekman** [comment]
relevant to https://github.com/Homebrew/homebrew/pull/46547

**Comment by afanfa** [comment]
Hi, I cannot reproduce the issue with allocate_as_brier_proc. I'm using gcc-5.2, opencoarrays-1.2.0, mpich-3.1.4, cmake-3.1.3 on Ubuntu. Can you post the error you get?

**Comment by zbeekman** [comment]
OK let me see what's going on. I was about to say with confidence that it happens with OpenMPI on Mac OS X all the time, however, I just tried it with OpenMPI (and clang for mpicc) and it didn't happen. It was happening with gcc 5.2 for mpicc when building on the travis-ci homebrew test servers. Let me dig around some more and get back to you.

**Comment by zbeekman** [comment]
this is maddening.... it seems that there is something from my environment that allows the build and test (by hand) with brewed GCC 5.2, Clang, OpenMPI to work, but when building and testing under the restricted environment of Homebrew this doesn't work.

The output is just simply: "Test failed." It doesn't give any reason. No segfaults or anything like that.

**Comment by zbeekman** [comment]
OK upon looking at the definition of the unit test, and reading the output more carefully, it says:
```
test 6
      Start  6: allocate_as_barrier_proc

6: Test command: /usr/local/bin/mpiexec "-np" "2" "/tmp/opencoarrays20151202-44375-14o6csk/opencoarrays/build/src/tests/unit/init_register/allocate_as_barrier_proc"
6: Test timeout computed to be: 9.99988e+06
6:  Test failed.           1
 6/21 Test  #6: allocate_as_barrier_proc .........***Failed  Required regular expression not found.Regex=[Test passed.
]  1.07 sec
```

So [here](https://github.com/sourceryinstitute/opencoarrays/blob/master/src/tests/unit/init_register/allocate_as_barrier_proc.f90#L11) `me /= 1` but on the following line it does equal 1...

If my memory is correct, `deallocate` should act as an image control statement/barrier too, shouldn't it? It seems that, perhaps, this is the problem, since the value of `me` changes from the `if` statement to the `print` statement.

**Comment by afanfa** [comment]
Inside the directory chosen for installing opencoarrays with CMake, there a directory called Temporary. Inside that directory the file called LastTest.log reports the output of all the tests. Can you post the output of allocate_as_barrier_proc?

**Comment by zbeekman** [comment]
When running `test -E co_reduce` this is LastTest.log:
```
Start testing: Dec 02 13:56 EST
----------------------------------------------------------
1/20 Testing: initialize_mpi
1/20 Test: initialize_mpi
Command: "/usr/local/bin/mpiexec" "-np" "2" "/tmp/opencoarrays20151202-56450-hbtpjn/opencoarrays/build/src/tests/unit/init_register/initialize_mpi"
Directory: /tmp/opencoarrays20151202-56450-hbtpjn/opencoarrays/build
"initialize_mpi" start time: Dec 02 13:56 EST
Output:
----------------------------------------------------------
 Test passed.
<end of output>
Test time =   0.06 sec
----------------------------------------------------------
Test Pass Reason:
Required regular expression found.Regex=[Test passed.
]
"initialize_mpi" end time: Dec 02 13:56 EST
"initialize_mpi" time elapsed: 00:00:00
----------------------------------------------------------

2/20 Testing: register
2/20 Test: register
Command: "/usr/local/bin/mpiexec" "-np" "2" "/tmp/opencoarrays20151202-56450-hbtpjn/opencoarrays/build/src/tests/unit/init_register/register"
Directory: /tmp/opencoarrays20151202-56450-hbtpjn/opencoarrays/build
"register" start time: Dec 02 13:56 EST
Output:
----------------------------------------------------------
 Test passed.
<end of output>
Test time =   0.06 sec
----------------------------------------------------------
Test Pass Reason:
Required regular expression found.Regex=[Test passed.
]
"register" end time: Dec 02 13:56 EST
"register" time elapsed: 00:00:00
----------------------------------------------------------

3/20 Testing: register_rename_me
3/20 Test: register_rename_me
Command: "/usr/local/bin/mpiexec" "-np" "2" "/tmp/opencoarrays20151202-56450-hbtpjn/opencoarrays/build/src/tests/unit/init_register/register_rename_me"
Directory: /tmp/opencoarrays20151202-56450-hbtpjn/opencoarrays/build
"register_rename_me" start time: Dec 02 13:56 EST
Output:
----------------------------------------------------------
 Test passed.
<end of output>
Test time =   0.07 sec
----------------------------------------------------------
Test Pass Reason:
Required regular expression found.Regex=[Test passed.
]
"register_rename_me" end time: Dec 02 13:56 EST
"register_rename_me" time elapsed: 00:00:00
----------------------------------------------------------

4/20 Testing: register_rename_me_too
4/20 Test: register_rename_me_too
Command: "/usr/local/bin/mpiexec" "-np" "2" "/tmp/opencoarrays20151202-56450-hbtpjn/opencoarrays/build/src/tests/unit/init_register/register_rename_me_too"
Directory: /tmp/opencoarrays20151202-56450-hbtpjn/opencoarrays/build
"register_rename_me_too" start time: Dec 02 13:56 EST
Output:
----------------------------------------------------------
 Test passed.
<end of output>
Test time =   0.06 sec
----------------------------------------------------------
Test Pass Reason:
Required regular expression found.Regex=[Test passed.
]
"register_rename_me_too" end time: Dec 02 13:56 EST
"register_rename_me_too" time elapsed: 00:00:00
----------------------------------------------------------

5/20 Testing: allocate_as_barrier
5/20 Test: allocate_as_barrier
Command: "/usr/local/bin/mpiexec" "-np" "2" "/tmp/opencoarrays20151202-56450-hbtpjn/opencoarrays/build/src/tests/unit/init_register/allocate_as_barrier"
Directory: /tmp/opencoarrays20151202-56450-hbtpjn/opencoarrays/build
"allocate_as_barrier" start time: Dec 02 13:56 EST
Output:
----------------------------------------------------------
 Test passed.
<end of output>
Test time =   1.08 sec
----------------------------------------------------------
Test Pass Reason:
Required regular expression found.Regex=[Test passed.
]
"allocate_as_barrier" end time: Dec 02 13:56 EST
"allocate_as_barrier" time elapsed: 00:00:01
----------------------------------------------------------

6/20 Testing: allocate_as_barrier_proc
6/20 Test: allocate_as_barrier_proc
Command: "/usr/local/bin/mpiexec" "-np" "2" "/tmp/opencoarrays20151202-56450-hbtpjn/opencoarrays/build/src/tests/unit/init_register/allocate_as_barrier_proc"
Directory: /tmp/opencoarrays20151202-56450-hbtpjn/opencoarrays/build
"allocate_as_barrier_proc" start time: Dec 02 13:56 EST
Output:
----------------------------------------------------------
 Test failed.           1
<end of output>
Test time =   1.06 sec
----------------------------------------------------------
Test Fail Reason:
Required regular expression not found.Regex=[Test passed.
]
"allocate_as_barrier_proc" end time: Dec 02 13:56 EST
"allocate_as_barrier_proc" time elapsed: 00:00:01
----------------------------------------------------------

7/20 Testing: get_array
7/20 Test: get_array
Command: "/usr/local/bin/mpiexec" "-np" "2" "/tmp/opencoarrays20151202-56450-hbtpjn/opencoarrays/build/src/tests/unit/send-get/get_array"
Directory: /tmp/opencoarrays20151202-56450-hbtpjn/opencoarrays/build
"get_array" start time: Dec 02 13:56 EST
Output:
----------------------------------------------------------
 Test passed
 Test passed
<end of output>
Test time =   0.10 sec
----------------------------------------------------------
Test Pass Reason:
Required regular expression found.Regex=[Test passed.
]
"get_array" end time: Dec 02 13:56 EST
"get_array" time elapsed: 00:00:00
----------------------------------------------------------

8/20 Testing: send_array
8/20 Test: send_array
Command: "/usr/local/bin/mpiexec" "-np" "2" "/tmp/opencoarrays20151202-56450-hbtpjn/opencoarrays/build/src/tests/unit/send-get/send_array"
Directory: /tmp/opencoarrays20151202-56450-hbtpjn/opencoarrays/build
"send_array" start time: Dec 02 13:56 EST
Output:
----------------------------------------------------------
 Test passed
 Test passed
<end of output>
Test time =   0.11 sec
----------------------------------------------------------
Test Pass Reason:
Required regular expression found.Regex=[Test passed.
]
"send_array" end time: Dec 02 13:56 EST
"send_array" time elapsed: 00:00:00
----------------------------------------------------------

9/20 Testing: get_with_offset_1d
9/20 Test: get_with_offset_1d
Command: "/usr/local/bin/mpiexec" "-np" "2" "/tmp/opencoarrays20151202-56450-hbtpjn/opencoarrays/build/src/tests/unit/send-get/get_with_offset_1d"
Directory: /tmp/opencoarrays20151202-56450-hbtpjn/opencoarrays/build
"get_with_offset_1d" start time: Dec 02 13:56 EST
Output:
----------------------------------------------------------
 Test passed.
<end of output>
Test time =   0.06 sec
----------------------------------------------------------
Test Pass Reason:
Required regular expression found.Regex=[Test passed.
]
"get_with_offset_1d" end time: Dec 02 13:56 EST
"get_with_offset_1d" time elapsed: 00:00:00
----------------------------------------------------------

10/20 Testing: whole_get_array
10/20 Test: whole_get_array
Command: "/usr/local/bin/mpiexec" "-np" "2" "/tmp/opencoarrays20151202-56450-hbtpjn/opencoarrays/build/src/tests/unit/send-get/whole_get_array"
Directory: /tmp/opencoarrays20151202-56450-hbtpjn/opencoarrays/build
"whole_get_array" start time: Dec 02 13:56 EST
Output:
----------------------------------------------------------
 Test passed.
<end of output>
Test time =   0.07 sec
----------------------------------------------------------
Test Pass Reason:
Required regular expression found.Regex=[Test passed.
]
"whole_get_array" end time: Dec 02 13:56 EST
"whole_get_array" time elapsed: 00:00:00
----------------------------------------------------------

11/20 Testing: strided_get
11/20 Test: strided_get
Command: "/usr/local/bin/mpiexec" "-np" "2" "/tmp/opencoarrays20151202-56450-hbtpjn/opencoarrays/build/src/tests/unit/send-get/strided_get"
Directory: /tmp/opencoarrays20151202-56450-hbtpjn/opencoarrays/build
"strided_get" start time: Dec 02 13:56 EST
Output:
----------------------------------------------------------
 Test passed.
<end of output>
Test time =   0.06 sec
----------------------------------------------------------
Test Pass Reason:
Required regular expression found.Regex=[Test passed.
]
"strided_get" end time: Dec 02 13:56 EST
"strided_get" time elapsed: 00:00:00
----------------------------------------------------------

12/20 Testing: co_sum
12/20 Test: co_sum
Command: "/usr/local/bin/mpiexec" "-np" "4" "/tmp/opencoarrays20151202-56450-hbtpjn/opencoarrays/build/src/tests/unit/collectives/co_sum_test"
Directory: /tmp/opencoarrays20151202-56450-hbtpjn/opencoarrays/build
"co_sum" start time: Dec 02 13:56 EST
Output:
----------------------------------------------------------
 Test passed.
<end of output>
Test time =   0.07 sec
----------------------------------------------------------
Test Pass Reason:
Required regular expression found.Regex=[Test passed.
]
"co_sum" end time: Dec 02 13:56 EST
"co_sum" time elapsed: 00:00:00
----------------------------------------------------------

13/20 Testing: co_broadcast
13/20 Test: co_broadcast
Command: "/usr/local/bin/mpiexec" "-np" "4" "/tmp/opencoarrays20151202-56450-hbtpjn/opencoarrays/build/src/tests/unit/collectives/co_broadcast_test"
Directory: /tmp/opencoarrays20151202-56450-hbtpjn/opencoarrays/build
"co_broadcast" start time: Dec 02 13:56 EST
Output:
----------------------------------------------------------
 Test passed.
<end of output>
Test time =   0.07 sec
----------------------------------------------------------
Test Pass Reason:
Required regular expression found.Regex=[Test passed.
]
"co_broadcast" end time: Dec 02 13:56 EST
"co_broadcast" time elapsed: 00:00:00
----------------------------------------------------------

14/20 Testing: co_min
14/20 Test: co_min
Command: "/usr/local/bin/mpiexec" "-np" "4" "/tmp/opencoarrays20151202-56450-hbtpjn/opencoarrays/build/src/tests/unit/collectives/co_min_test"
Directory: /tmp/opencoarrays20151202-56450-hbtpjn/opencoarrays/build
"co_min" start time: Dec 02 13:56 EST
Output:
----------------------------------------------------------
 Test passed.
<end of output>
Test time =   0.08 sec
----------------------------------------------------------
Test Pass Reason:
Required regular expression found.Regex=[Test passed.
]
"co_min" end time: Dec 02 13:56 EST
"co_min" time elapsed: 00:00:00
----------------------------------------------------------

15/20 Testing: co_max
15/20 Test: co_max
Command: "/usr/local/bin/mpiexec" "-np" "4" "/tmp/opencoarrays20151202-56450-hbtpjn/opencoarrays/build/src/tests/unit/collectives/co_max_test"
Directory: /tmp/opencoarrays20151202-56450-hbtpjn/opencoarrays/build
"co_max" start time: Dec 02 13:56 EST
Output:
----------------------------------------------------------
 Correct real(c_double) co_max
 Test passed.
<end of output>
Test time =   0.07 sec
----------------------------------------------------------
Test Pass Reason:
Required regular expression found.Regex=[Test passed.
]
"co_max" end time: Dec 02 13:56 EST
"co_max" time elapsed: 00:00:00
----------------------------------------------------------

16/20 Testing: syncall
16/20 Test: syncall
Command: "/usr/local/bin/mpiexec" "-np" "32" "/tmp/opencoarrays20151202-56450-hbtpjn/opencoarrays/build/src/tests/unit/sync/syncall"
Directory: /tmp/opencoarrays20151202-56450-hbtpjn/opencoarrays/build
"syncall" start time: Dec 02 13:56 EST
Output:
----------------------------------------------------------
 Test passed.
--------------------------------------------------------------------------
A system call failed during shared memory initialization that should
not have.  It is likely that your MPI job will now either abort or
experience performance degradation.

  Local host:  IBB-MBP.local
  System call: unlink(2) /var/folders/4t/my3vyzsx7pg80tc_cxhyfc7h0000gp/T/openmpi-sessions-ibeekman@IBB-MBP_0/8537/1/shared_window_4.IBB-MBP
  Error:       No such file or directory (errno 2)
--------------------------------------------------------------------------
<end of output>
Test time =   1.32 sec
----------------------------------------------------------
Test Pass Reason:
Required regular expression found.Regex=[Test passed.
]
"syncall" end time: Dec 02 13:56 EST
"syncall" time elapsed: 00:00:01
----------------------------------------------------------

17/20 Testing: syncimages
17/20 Test: syncimages
Command: "/usr/local/bin/mpiexec" "-np" "32" "/tmp/opencoarrays20151202-56450-hbtpjn/opencoarrays/build/src/tests/unit/sync/syncimages"
Directory: /tmp/opencoarrays20151202-56450-hbtpjn/opencoarrays/build
"syncimages" start time: Dec 02 13:56 EST
Output:
----------------------------------------------------------
 Test passed.
<end of output>
Test time =   0.30 sec
----------------------------------------------------------
Test Pass Reason:
Required regular expression found.Regex=[Test passed.
]
"syncimages" end time: Dec 02 13:56 EST
"syncimages" time elapsed: 00:00:00
----------------------------------------------------------

18/20 Testing: hello_multiverse
18/20 Test: hello_multiverse
Command: "/usr/local/bin/mpiexec" "-np" "2" "/tmp/opencoarrays20151202-56450-hbtpjn/opencoarrays/build/src/tests/integration/coarrayHelloWorld/hello_multiverse"
Directory: /tmp/opencoarrays20151202-56450-hbtpjn/opencoarrays/build
"hello_multiverse" start time: Dec 02 13:56 EST
Output:
----------------------------------------------------------
 Greetings from image  1 of  2                                                                       
 Greetings from image  2 of  2                                                                       
 Test passed.
<end of output>
Test time =   0.06 sec
----------------------------------------------------------
Test Pass Reason:
Required regular expression found.Regex=[Test passed.
]
"hello_multiverse" end time: Dec 02 13:56 EST
"hello_multiverse" time elapsed: 00:00:00
----------------------------------------------------------

19/20 Testing: coarray_burgers_pde
19/20 Test: coarray_burgers_pde
Command: "/usr/local/bin/mpiexec" "-np" "2" "/tmp/opencoarrays20151202-56450-hbtpjn/opencoarrays/build/src/tests/integration/pde_solvers/coarrayBurgers/coarray_burgers_pde"
Directory: /tmp/opencoarrays20151202-56450-hbtpjn/opencoarrays/build
"coarray_burgers_pde" start time: Dec 02 13:56 EST
Output:
----------------------------------------------------------
 Time =  0.60065795534754485     
 On image            1 u =  -3.6062644769485318E-016  0.55703975723770405        1.1137114029330639        1.6695096165091936        2.2228610006371610        2.7616422370249762        3.1602423341809680        2.5585980904689771     
 On image            2 u =   1.9794875987846594E-015  -2.5585980904689749       -3.1602423341809680       -2.7616422370249771       -2.2228610006371614       -1.6695096165091943       -1.1137114029330646      -0.55703975723770482     
 Test passed.
<end of output>
Test time =   0.07 sec
----------------------------------------------------------
Test Pass Reason:
Required regular expression found.Regex=[Test passed.
]
"coarray_burgers_pde" end time: Dec 02 13:56 EST
"coarray_burgers_pde" time elapsed: 00:00:00
----------------------------------------------------------

20/20 Testing: co_heat
20/20 Test: co_heat
Command: "/usr/local/bin/mpiexec" "-np" "2" "/tmp/opencoarrays20151202-56450-hbtpjn/opencoarrays/build/src/tests/integration/pde_solvers/coarrayHeatSimplified/co_heat"
Directory: /tmp/opencoarrays20151202-56450-hbtpjn/opencoarrays/build
"co_heat" start time: Dec 02 13:56 EST
Output:
----------------------------------------------------------
 Test passed.
<end of output>
Test time =   0.74 sec
----------------------------------------------------------
Test Pass Reason:
Required regular expression found.Regex=[Test passed.
]
"co_heat" end time: Dec 02 13:56 EST
"co_heat" time elapsed: 00:00:00
----------------------------------------------------------

End testing: Dec 02 13:56 EST
```

**Comment by afanfa** [comment]
Got it. For reasons that I don't know, your environment is using an optimization flag different from -O0. There is a known issues with GFortran about this phenomena (https://groups.google.com/forum/#!msg/opencoarrays/RZOwwYTqG80/46S9eL696dgJ).

**Comment by zbeekman** [comment]
OK, so to clarify, if I'm reading the email thread correctly:

There is a gcc/gfortran bug that causes -O2 optimization to give incorrect results. This problem will be fixed in an upcoming gcc release.

Does the advice about optimization flag apply only to compiling the tests, or also the entire opencoarrays package?

**Comment by zbeekman** [comment]
Also, this might be worth adding to the STATUS.md

**Comment by zbeekman** [comment]
I suspect the cause of this is passing `*std_cmake_args` to CMake from the homebrew formula. This results in:

```
cmake .. -DCMAKE_C_FLAGS_RELEASE= -DCMAKE_CXX_FLAGS_RELEASE= -DCMAKE_INSTALL_PREFIX=/usr/local/homebrew/Cellar/opencoarrays/1.2.0 -DCMAKE_BUILD_TYPE=Release -DCMAKE_FIND_FRAMEWORK=LAST -DCMAKE_VERBOSE_MAKEFILE=ON -Wno-dev
```

The culprit is `-DCMAKE_BUILD_TYPE=Release` I just confirmed this. I think that the CMakeLists.txt should probably be set up to function properly even when a "Release" build type is selected... Maybe I'll file a new issue or submit a PR for that. Thanks for helping me get to the bottom of this, @afanfa, I really appreciate it!

**Comment by afanfa** [comment]
This issue has been in my todo list for too long. It is time to fix it. Thanks for working on this.

**Comment by zbeekman** [comment]
PR #43 should take care of this issue, and also update the STATUS.md. PR made against master, so commits will probably need to be [`git cherry-pick`ed](https://git-scm.com/docs/git-cherry-pick) into the `branch-1.0.0` branch.

**Comment by zbeekman** [comment]
Prematurely closed this before actually fixing the CMakeLists.txt... reopening.

**Comment by zbeekman** [comment]
The bug in question has been patched in GCC 5.3 and trunk: https://groups.google.com/forum/#!topic/opencoarrays/038m9fZP2GQ
